src/ngraph/pass/constant_folding.cpp

```diff
@@ -365,78 +381,227 @@ void pass::ConstantFolding::construct_constant_broadcast()
         broadcast_matcher, constant_broadcast_callback, PassProperty::REQUIRE_STATIC_SHAPE);
 }
 
-template <class T>
+template <class Tin, class Tout>
 shared_ptr<op::Constant> fold_constant_binary(shared_ptr<op::Constant> a,
                                               shared_ptr<op::Constant> b,
                                               shared_ptr<Node> binary,
                                               NodeExecutorTy func)
 {
     auto out_shape = binary->get_shape();
-    vector<T> out_vec(shape_size(out_shape));
 
     if (func != nullptr)
     {
+        vector<Tout> out_vec(shape_size(out_shape));
         vector<void*> inputs;
         inputs.push_back(const_cast<void*>(a->get_data_ptr()));
         inputs.push_back(const_cast<void*>(b->get_data_ptr()));
         vector<void*> outputs;
         outputs.push_back(out_vec.data());
 
         func(inputs, outputs);
+        return make_shared<op::Constant>(binary->get_element_type(), out_shape, out_vec);
     }
     else
     {
         if (std::dynamic_pointer_cast<op::Add>(binary))
         {
-            runtime::reference::add<T>(
-                a->get_data_ptr<T>(), b->get_data_ptr<T>(), out_vec.data(), shape_size(out_shape));
-        }
-        else if (std::dynamic_pointer_cast<op::Subtract>(binary))
-        {
-            runtime::reference::subtract<T>(
-                a->get_data_ptr<T>(), b->get_data_ptr<T>(), out_vec.data(), shape_size(out_shape));
+            NGRAPH_CHECK(element::from<Tin>() == element::from<Tout>(),
+                         "Input/output types do not match");
+            vector<Tin> out_vec(shape_size(out_shape));
+            runtime::reference::add<Tin>(a->get_data_ptr<Tin>(),
+                                         b->get_data_ptr<Tin>(),
+                                         out_vec.data(),
+                                         shape_size(out_shape));
+            return make_shared<op::Constant>(binary->get_element_type(), out_shape, out_vec);
```

Comment:
Not a strong request, but there is a lot of code duplication here. Can we factor out setting the out_vec and initializing the Constant Op ? Probably better to have a type_id->reference_kernel map also.

Reply:
Agreed that the code duplication here is getting kind of hideous. Will see what I can factor out in a follow-on PR. I have some ideas for a broader refactor of the reference kernels, but would like to save that for an even later PR because it's a bit far-reaching.